### PR TITLE
aioredis v2 support

### LINKED ISSUE #648: aioredis 2.0 support
Add aioredis 2.0 support for [RedisStorage2](https://github.com/aiogram/aiogram/blob/dev-2.x/aiogram/contrib/fsm_storage/redis.py#L207)
https://aioredis.readthedocs.io/en/latest/migration/

### PR DESCRIPTION
**Pull request by Olegt0rr**
# Description
We should add non-compatibility-breakable support for aioredis v2

Fixes #648

## Type of change

Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by evgfilim1 — COMMENTED** [review]
IMO it breaks backwards compatibility anyway. `RedisStorage2.redis()` is public, so some code might rely on this method and its return type.
Also is it safe to rely on `importlib.metadata.version`?

**Comment by Olegt0rr** [comment]
> IMO it breaks backwards compatibility anyway. `RedisStorage2.redis()` is public, so some code might rely on this method and its return type.
> Also is it safe to rely on `importlib.metadata.version()`?

I've proxied methods from Redis, so now public `.redis()` method may be used as an original Redis

File: aiogram/contrib/fsm_storage/redis.py [hunk 1]
```diff
@@ -244,49 +386,59 @@ def __init__(self, host: str = 'localhost', port=6379, db=None, password=None,
...
+                db=self._db,
+                password=self._password,
+                ssl=self._ssl,
+                pool_size=self._pool_size,
+                loop=self._loop,
+                **self._kwargs,
+            )
+            if redis_version == 1:
+                self._redis = AioRedisAdapterV1(**connection_data)
+            elif redis_version == 2:
```
Comment (evgfilim1) [inline aiogram/contrib/fsm_storage/redis.py:414]: You're comparing `str` with `int`
```suggestion
            if redis_version == "1":
                self._redis = AioRedisAdapterV1(**connection_data)
            elif redis_version == "2":
```

File: aiogram/contrib/fsm_storage/redis.py [hunk 2]
```diff
@@ -244,49 +386,59 @@ def __init__(self, host: str = 'localhost', port=6379, db=None, password=None,
...
-                                                               db=self._db, password=self._password, ssl=self._ssl,
-                                                               minsize=1, maxsize=self._pool_size,
-                                                               loop=self._loop, **self._kwargs)
+        adapter = await self._get_adapter()
+        return await adapter.get_redis()
+
+    async def _get_adapter(self) -> AioRedisAdapterBase:
+        """Get adapter based on aioredis version."""
+        if self._redis is None:
+            redis_version = aioredis.__version__.split(".")[0]
```
Comment (evgfilim1) [inline aiogram/contrib/fsm_storage/redis.py:401]: See above
```suggestion
            redis_version = int(aioredis.__version__.split(".")[0])
```

**Review by evgfilim1 — CHANGES_REQUESTED** [review]
There's a bug. Choose one of two possible fixes

**Review by evgfilim1 — APPROVED** [review]

**Review by JrooTJunior — APPROVED** [review]

**Comment by paul1319** [comment]
There is a bug `AttributeError: module 'aioredis' has no attribute 'RedisConnection'`

File: aiogram/contrib/fsm_storage/redis.py [hunk 3]
```diff
@@ -204,6 +208,134 @@ async def update_bucket(self, *, chat: typing.Union[str, int, None] = None,
...
+                    ssl=self._ssl,
+                    minsize=1,
+                    maxsize=self._pool_size,
+                    loop=self._loop,
+                    **self._kwargs,
+                )
+        return self._redis
+
+    def close(self):
+        async with self._connection_lock:
```
[the remark this answers is not in the document]
Comment (Olegt0rr) [reply aiogram/contrib/fsm_storage/redis.py:299]: Fixed in #676

**Pull request merged**